cloud/pkg/upgradecontroller/controller/util.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2022 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"strings"
+
+	metav1 "k8s.io/api/core/v1"
+
+	"github.com/kubeedge/kubeedge/pkg/apis/upgrade/v1alpha2"
+)
+
+// filterVersion returns true only if the edge node version already on the upgrade req
+// version is like: v1.22.6-kubeedge-v1.10.0-beta.0.185+95378fb019912a, expected is like v1.10.0
+func filterVersion(version string, expected string) bool {
+	// if not correct version format, also return true
+	index := strings.Index(version, "-kubeedge-")
+	if index == -1 {
+		return false
+	}
+
+	if version[index+10:] != expected {
+		// filter nodes that already in the required version
+		return false
+	}
+
+	return true
+}
+
+// isEdgeNode checks whether a node is an Edge Node
+// only if label {"node-role.kubernetes.io/edge": ""} exists, it is an edge node
+func isEdgeNode(node *metav1.Node) bool {
+	if node.Labels == nil {
+		return false
+	}
+	if _, ok := node.Labels["node-role.kubernetes.io/edge"]; !ok {
+		return false
+	}
+
+	if node.Labels["node-role.kubernetes.io/edge"] != "" {
+		return false
+	}
+
+	return true
+}
+
+// isCompleted returns true only if ReUpgrade field is true and some/all edge upgrade is upgrading or completed
+func isCompleted(upgrade *v1alpha2.Upgrade) bool {
+	if !upgrade.Spec.ReUpgrade {
+		// all edge node upgrade is upgrading or completed
```

Comment:
I am a little confusing about this field ReUpgrade, correcting me if i am wrong, the controller cannot handle the reupgrade steps?

Reply:
if the former upgrade failed, we can retrigger the upgrade by setting this field to true. If this field is false, we'll not send upgrade msg to edge again if the former upgrade is completed. my proposal is one upgrade cr for one version upgrade.

Reply:
> if the former upgrade failed, we can retrigger the upgrade by setting this field to true. If this field is false, we'll not send upgrade msg to edge again if the former upgrade is completed. my proposal is one upgrade cr for one version upgrade.

Got it.  If an upgrade step holding or failed, we had better have a retry I think.

Reply:
concerned about causing data loss if we retry, so if the upgrade failed, users may need to check why it failed or retrigger again by hand. So once a cr is created, we'll only trigger upgrade once a time

Reply:
How about designing a ReTry struct as a optional for users to design it. The logs also need to store in history.